python/ray/worker.py

```diff
@@ -468,6 +468,21 @@ def get_object(self, object_ids):
 
         if len(unready_ids) > 0:
             with self.state_lock:
+                # Get the task ID, to notify the backend which task is blocked.
+                current_task_id = self.current_task_id
+                if not ray.utils.is_main_thread():
+                    # If this is running on a separate thread, then the mapping
+                    # to the current task ID may not be correct. Generate a
+                    # random task ID so that the backend can differentiate
+                    # between different threads.
+                    current_task_id = ray.ObjectID(random_string())
+                    if not self.multithreading_warned:
+                        logger.warning(
+                            "Calling ray.get or ray.wait in a separate thread "
+                            "may lead to deadlock")
```

Comment:
How do you get deadlock?

Reply:
It can happen if the main thread is blocked indirectly on a `ray.get`. For instance, the main thread spawns some other threads that all block in a `ray.get`, and then the main thread joins the other threads. Let me know if you can think of a better message to print to the user for this...

Reply:
Hm I don't write get that still, doesn't it still work out once the gets complete?

Reply:
Ah oops, forgot to mention that the issue is that normally we reclaim the resources that a worker is using if it's blocked in a `ray.get`, but we don't do it if it's a separate thread on the worker is blocked. So if there is only 1 CPU, the worker will use it and won't give up its resources while waiting for the threads to finish. Then, the `ray.get`s will never return because there are no other resources available to execute the needed tasks. This is why I changed the `test_multithreading` test case to use 2 CPUs.

Reply:
Ah I see, then I think the warning can say something about not yielding resources, which can lead to deadlock. The current warning makes it sound like it can happen spuriously.

Reply:
So another option: could we consider the worker blocked if *any* thread is doing a ray.get or wait? That way, no such deadlock is possible.

Reply:
Hmm I worry about doing that since the worker will not actually be blocked and it will be easy to oversubscribe the resources.